### ISSUE TITLE
Uncaught exception/crash in parser with panic: strconv.Atoi - Invalid Syntax when a variable named "config" (or any other variable name which may duplicate used HCL stanza keys) is present in remote state

### ISSUE THREAD

**Issue by scianos** [hashicorp/terraform#19504]
### Terraform Version

```
Terraform v0.11.10
+ provider.archive v1.1.0
+ provider.aws v1.49.0
+ provider.null v1.0.0
+ provider.template v1.0.0
```

### Terraform Configuration Files

```hcl
terraform {
  backend "s3" {
    bucket = "org-tfstate-1234567890"
    key    = "env-e1/project_state_test"
    region = "us-west-1"
  }
}

data "terraform_remote_state" "account_baseline" {
  backend = "s3"

  config {
    bucket = "org-tfstate-1234567890"
    key    = "env-e1/account_baseline"
    region = "us-west-1"
  }
}

data "terraform_remote_state" "project_s3_buckets" {
  backend = "s3"

  config {
    bucket = "org-tfstate-1234567890"
    key    = "env-e1/project_s3_buckets"
    region = "us-west-1"
  }
}
provider "aws" {
  region  = "us-west-1"
  version = "~> 1.9"
}

data "aws_caller_identity" "current" {}

module "defaults" {
  source = "../1_defaults"
}

output "managed_buckets" {
  value = "${values(data.terraform_remote_state.s3_buckets.buckets)}"
}

output "unmanaged_buckets" {
  value = "${values(data.terraform_remote_state.s3_buckets.unmanaged_buckets)}"
}
```

### Debug Output

https://gist.github.com/scianos/605730c80c0367a2d631fa92c38944bf

### Crash Output

https://gist.github.com/scianos/96575e54d1d987a0b5d6ad2ee0d5b32d

### Expected Behavior
The config variables should not be conflated in both in-memory state representation nor the persisted output across namespaces, or, if Terraform requires global unique names across modules or has the concept of reserved words, ensure proper error trapping within the parser.

### Actual Behavior
Configuration output is conflated between namespaces in the state, resulting in parser failure. To wit, the persisted state shows:
```
                            "config.#": "1",  <-- THE PROVIDER "CONFIG" VARIABLE
                            "config.%": "13", <-- THE "CONFIG" VARIABLE PULLED FROM REMOTE STATE/MODULE OUTPUT
                            "config.987654321.bucket": "org-tfstate-1234567890",
                            "config.987654321.key": "env-e1/project_s3_buckets",
                            "config.987654321.region": "us-west-1",
                            "config.org-log-application-abc-all-1234567890.%": "4",
                            "config.org-log-application-abc-all-1234567890.environment": "all",
                            "config.org-log-application-abc-all-1234567890.function": "application",
```

### Steps to Reproduce
 1. Store a variable called "config" into remote state
 2. Bring in above remote state into another module using the AWS provider with state stored in an S3 bucket, and relevant configuration parameters for the bucket/key present in the module definition.
 3. Execute terraform plan/apply
 4. If Terraform does not immediately crash, pull the state and check for conflation of config variables across namespaces as described above under Actual Behavior.
 5. If Terraform does crash, validate this was the issue by renaming the config variable and re-attempting apply.

### Additional Context
The immediate red flag which caused me to look at the parser as well as possible conflation of state for entities with duplicate names across the namespace is this piece of code, and the specific branch which executed according to trace logs:
from terraform/interpolate.go
```
714 func (i *Interpolater) interpolateComplexTypeAttribute(
715         resourceID string,
716         attributes map[string]string) (ast.Variable, error) {
717         // We can now distinguish between lists and maps in state by the count field:
718         //    - lists (and by extension, sets) use the traditional .# notation
719         //    - maps use the newer .% notation
720         // Consequently here we can decide how to deal with the keys appropriately
721         // based on whether the type is a map of list.
722         if lengthAttr, isList := attributes[resourceID+".#"]; isList {
723                 log.Printf("[DEBUG] Interpolating computed list element attribute %s (%s)",
724                         resourceID, lengthAttr)
725
726                 // In Terraform's internal dotted representation of list-like attributes, the
727                 // ".#" count field is marked as unknown to indicate "this whole list is
728                 // unknown". We must honor that meaning here so computed references can be
729                 // treated properly during the plan phase.
730                 if lengthAttr == config.UnknownVariableValue {
731                         return unknownVariable(), nil
732                 }
733
734                 expanded := flatmap.Expand(attributes, resourceID)
735                 return hil.InterfaceToVariable(expanded)
736         }
737
738         if lengthAttr, isMap := attributes[resourceID+".%"]; isMap {
739                 log.Printf("[DEBUG] Interpolating computed map element attribute %s (%s)",
740                         resourceID, lengthAttr)
741
742                 // In Terraform's internal dotted representation of map attributes, the
743                 // ".%" count field is marked as unknown to indicate "this whole list is
744                 // unknown". We must honor that meaning here so computed references can be
745                 // treated properly during the plan phase.
746                 if lengthAttr == config.UnknownVariableValue {
747                         return unknownVariable(), nil
748                 }
749
750                 expanded := flatmap.Expand(attributes, resourceID)
751                 return hil.InterfaceToVariable(expanded)
752         }
753
754         return ast.Variable{}, fmt.Errorf("No complex type %s found", resourceID)
755 }
```
Above, we see that a postfix notation is used to identify if a resource is a map or a list. During testing, we noted that the branch handling list expansion was being used to parse values which should have been expanded as a map. We then began examining the state after each run of Terraform and noted corruption of the order of the state file output for the provider's "config" map in state and our module's "config" map/output.

Please note that we believe this issue _may_ have been introduced around 0.11.7 or greater; we did not reproduce on earlier versions.

**Comment by apparentlymart** [comment]
The codepath where this crash comes from is no longer used as we prepare the master branch for the forthcoming v0.12.0 release, so it is likely that this problem is already fixed by the changes both to the expression evaluation and to the remote state data source itself.

I'm going to label this to prompt us to verify the fix before final release.

**Comment by scianos** [comment]
Thanks much for the update @apparentlymart - we too believe this is fixed in v12 during initial testing of TF v12 alpha's ability to parse the config without error (though we chose not to migrate to v12 in its alpha state for production usage). In the meantime, we have worked around the issue by avoiding the use of the reserved words as variable names and hopefully this bug report will serve as documentation of workaround for any other folks bumping into this.

TL;DR - for anyone running into this problem, avoid the use of Terraform HCL configuration keys as variable names/outputs in versions below 12.

**Comment by apparentlymart** [comment]
Hi again @scianos!

I've now verified this with the 0.12.0-alpha4 release.

I created two configurations. My "producer" configuration just produces a simple output:

```hcl
output "config" {
  value = {
    a = 1
    b = "hello"
  }
}
```

I applied this to have it create a `terraform.tfstate` containing the given output, and then wrote a separate "consumer" configuration in a different directory:

```hcl
data "terraform_remote_state" "foo" {
  backend = "local"
  config = {
    path = "${path.module}/../producer/terraform.tfstate"
  }
}

output "config" {
  value = data.terraform_remote_state.foo.outputs.config
}
```

After applying this, I got the expected result:

```
$ terraform apply
data.terraform_remote_state.foo: Refreshing state...

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

config = {
  "a" = 1
  "b" = "hello"
}
```

Note that I wrote `data.terraform_remote_state.foo.outputs.config` instead of just `data.terraform_remote_state.foo.config` here. This new `outputs` key is how the output names from the remote state are kept separate from the arguments of the remote state data source itself. This can also be seen in the data source state:

```json
        {
          "schema_version": 0,
          "attributes": {
            "backend": "local",
            "config": {
              "value": {
                "path": "../producer/terraform.tfstate"
              },
              "type": [
                "object",
                {
                  "path": "string"
                }
              ]
            },
            "defaults": null,
            "outputs": {
              "value": {
                "config": {
                  "a": 1,
                  "b": "hello"
                }
              },
              "type": [
                "object",
                {
                  "config": [
                    "object",
                    {
                      "a": "number",
                      "b": "string"
                    }
                  ]
                }
              ]
            },
            "workspace": null
          }
        }
```

Terraform 0.12 no longer uses that "flatmap" structure you saw before, so the data structure is expanded out as more intuitive JSON representation. This also shows the special way `outputs` is populated where it includes both the values themselves and the type information from the source configuration, so that Terraform can ensure that these values will have exactly the same type in the consuming configuration, rather than type information being lost in the JSON serialization process (since the JSON type system is a subset of Terraform's).

---

Interestingly, this _particular_ situation behaves in an interesting way under an upgrade to 0.12 with your configuration example as given. Since `config` is defined as an argument for `terraform_remote_state`, accessing it without that extra `.outputs` key doesn't produce an error but instead returns the configuration:

```
$ terraform apply
data.terraform_remote_state.foo: Refreshing state...

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

config = {
  "path" = "../producer/terraform.tfstate"
}
```

Since such a configuration would've panicked under 0.11 I don't think this is a big concern, but definitely something to keep in mind during upgrade. The configuration upgrade tool is able to deal with inserting this extra `.outputs` key in most cases, but in this particular case it assumes you wanted to access the configuration of the backend itself and so it won't do any rewriting.

---

Since these changes are already in the master branch and ready to be included in the forthcoming v0.12 release, I'm going to close this out now. Thanks for reporting this, @scianos!
